GAE/src/test/java/org/akvo/flow/xml/FlowXmlObjectReaderTests.java

```diff
@@ -0,0 +1,212 @@
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *  See the GNU Affero General Public License included below for more details.
+ *
+ *  The full license text can also be seen at <http://www.gnu.org/licenses/agpl.html>.
+ */
+
+package test.java.org.akvo.flow.xml;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.io.IOException;
+
+import org.akvo.flow.xml.PublishedForm;
+import org.akvo.flow.xml.XmlForm;
+import org.junit.jupiter.api.Test;
+import org.waterforpeople.mapping.app.gwt.client.survey.QuestionDto;
+import org.waterforpeople.mapping.app.gwt.client.survey.QuestionDto.QuestionType;
+import org.waterforpeople.mapping.app.gwt.client.survey.QuestionGroupDto;
+import org.waterforpeople.mapping.app.gwt.client.survey.SurveyDto;
+
+class FlowXmlObjectReaderTests {
+
+    private String MINIMAL_XML_FORM = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>"
+            + "<survey name=\"Foo\" defaultLanguageCode=\"en\" version='1.0' app=\"akvoflowsandbox\" "
+            + "surveyGroupId=\"12345\" surveyGroupName=\"Bar\" surveyId=\"67890\">"
+            + "<questionGroup><heading>The Only Group</heading>"
+            + "<question order=\"1\" type=\"free\" mandatory=\"true\" localeNameFlag=\"false\" id=\"24680\">"
+            + "<altText type=\"translation\" language=\"sv\">Den enda frågan</altText>"
+            + "<text>The Only Question</text>"
+            + "</question></questionGroup></survey>";
+
+    private String QUESTIONLESS_XML_FORM = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>"
+            + "<survey name=\"Foo\" defaultLanguageCode=\"en\" version='1.0' app=\"akvoflowsandbox\" "
+            + "surveyGroupId=\"12345\" surveyGroupName=\"Bar\" surveyId=\"67890\">"
+            + "<questionGroup><heading>The Empty Group</heading>"
+            + "</questionGroup></survey>";
+
+    private String COMPATIBLE_XML_FORM = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>"
+            + "<survey name=\"Foo\" defaultLanguageCode=\"en\" version='1.0' app=\"akvoflowsandbox\" "
+            + "surveyGroupId=\"12345\" surveyGroupName=\"Bar\" surveyId=\"67890\">"
+            + "<altText type=\"translation\" language=\"sv\">Formuläret</altText>"
+            + "<questionGroup><heading>The Only Group</heading>"
+            + "<altText type=\"translation\" language=\"sv\">Den enda gruppen</altText>"
+            + "<question order=\"1\" type=\"free\" mandatory=\"true\" localeNameFlag=\"false\" id=\"24680\">"
+            + "<altText type=\"translation\" language=\"sv\">Den enda frågan</altText>"
+            + "<text>The Only Question</text>"
+            + "</question></questionGroup></survey>";
+
+    //Everything but the kitchen sink
```

Comment:
👍 a good set of samples.  Do you think its worth testing if the XML is erroneous or we shouldnt worry about that?

Reply:
I have already seen the test break on bad XML, so I am not worried about that...